### ISSUE TITLE
rar format: asan error for ./jtrts.pl -random -seed=987135 -type rar

### ISSUE THREAD

**Issue by frank-dittrich** [magnumripper/JohnTheRipper#1000]
```
(master)test $ OMP_NUM_THREADS=1 ./jtrts.pl -random -seed=987135 -stoponerror -type rar
-------------------------------------------------------------------------------
- JtR-TestSuite (jtrts). Version 1.13, Dec 21, 2014.  By, Jim Fougeron & others
- Testing:  John the Ripper password cracker, version 1.8.0.2-jumbo-1-bleeding_omp [linux-gnu 64-bit AVX2-autoconf]
--------------------------------------------------------------------------------

John Jumbo build detected.

form=rar                          guesses:    0 -show=  65 unk unk : Expected count(s) (130)  [!!!FAILED!!!  return code 256]
Exiting on error.  The pot file ./tst.pot contains the found data
The command used to run this test was:

../run/john -ses=./tst  -pot=./tst.pot rar_tst.in --wordlist=pw-40.dic -form=rar
```

Repeating this command:
```
(master)test $ rm tst.*
(master)test $ OMP_NUM_THREADS=1 ../run/john -ses=./tst  -pot=./tst.pot rar_tst.in --wordlist=pw-40.dic -form=rar
Loaded 130 password hashes with 130 different salts (rar, RAR3 [SHA1 AES 32/64])
Warning: OpenMP is disabled; a non-OpenMP build may be faster
Press 'q' or Ctrl-C to abort, almost any other key for status
hello            (test076.rar)
hello            (test036.rar)
hello            (test006.rar)
...
dffffffffffff    (test098.rar)
dffffffffffff    (test118.rar)
dffffffffffff    (test028.rar)
=================================================================
==10707== ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60040000d8d4 at pc 0x4ee9b7 bp 0x7fff053b0e20 sp 0x7fff053b0e10
READ of size 1 at 0x60040000d8d4 thread T0
    #0 0x4ee9b6 (/home/fd/git/JtR/run/john+0x4ee9b6)
    #1 0x4eeae8 (/home/fd/git/JtR/run/john+0x4eeae8)
    #2 0x4e24d4 (/home/fd/git/JtR/run/john+0x4e24d4)
    #3 0x4e7483 (/home/fd/git/JtR/run/john+0x4e7483)
    #4 0x6aeb48 (/home/fd/git/JtR/run/john+0x6aeb48)
    #5 0x6b040f (/home/fd/git/JtR/run/john+0x6b040f)
    #6 0x762978 (/home/fd/git/JtR/run/john+0x762978)
    #7 0x7644f5 (/home/fd/git/JtR/run/john+0x7644f5)
    #8 0x7646c0 (/home/fd/git/JtR/run/john+0x7646c0)
    #9 0x7b18da (/home/fd/git/JtR/run/john+0x7b18da)
    #10 0x777853 (/home/fd/git/JtR/run/john+0x777853)
    #11 0x7785b9 (/home/fd/git/JtR/run/john+0x7785b9)
    #12 0x38dd621d64 (/usr/lib64/libc-2.18.so+0x21d64)
    #13 0x4065b4 (/home/fd/git/JtR/run/john+0x4065b4)
0x60040000d8d4 is located 0 bytes to the right of 4-byte region [0x60040000d8d0,0x60040000d8d4)
allocated by thread T0 here:
    #0 0x7fe95965e219 (/usr/lib64/libasan.so.0.0.0+0x16219)
    #1 0x792293 (/home/fd/git/JtR/run/john+0x792293)
    #2 0x4de9ea (/home/fd/git/JtR/run/john+0x4de9ea)
    #3 0x4e63bd (/home/fd/git/JtR/run/john+0x4e63bd)
    #4 0x6aeb48 (/home/fd/git/JtR/run/john+0x6aeb48)
    #5 0x6b040f (/home/fd/git/JtR/run/john+0x6b040f)
    #6 0x762978 (/home/fd/git/JtR/run/john+0x762978)
    #7 0x7644f5 (/home/fd/git/JtR/run/john+0x7644f5)
    #8 0x7646c0 (/home/fd/git/JtR/run/john+0x7646c0)
    #9 0x7b18da (/home/fd/git/JtR/run/john+0x7b18da)
    #10 0x777853 (/home/fd/git/JtR/run/john+0x777853)
    #11 0x7785b9 (/home/fd/git/JtR/run/john+0x7785b9)
    #12 0x38dd621d64 (/usr/lib64/libc-2.18.so+0x21d64)
Shadow bytes around the buggy address:
  0x0c00ffff9ac0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9ad0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9ae0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9af0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9b00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c00ffff9b10: fa fa fa fa fa fa fa fa fa fa[04]fa fa fa fa fa
  0x0c00ffff9b20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9b30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c00ffff9b40: fa fa fa fa fa fa fa fa fa fa 00 00 fa fa fa fa
  0x0c00ffff9b50: fa fa fa fa fa fa fa fa fa fa 00 fa fa fa 00 fa
  0x0c00ffff9b60: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:     fa
  Heap righ redzone:     fb
  Freed Heap region:     fd
  Stack left redzone:    f1
  Stack mid redzone:     f2
  Stack right redzone:   f3
  Stack partial redzone: f4
  Stack after return:    f5
  Stack use after scope: f8
  Global redzone:        f9
  Global init order:     f6
  Poisoned by user:      f7
  ASan internal:         fe
==10707== ABORTING
```

```
(bleeding-jumbo)run $ ./asan_symbolize.py 
    #0 0x4ee9b6 (/home/fd/git/JtR/run/john+0x4ee9b6)
    #1 0x4eeae8 (/home/fd/git/JtR/run/john+0x4eeae8)
    #2 0x4e24d4 (/home/fd/git/JtR/run/john+0x4e24d4)
    #3 0x4e7483 (/home/fd/git/JtR/run/john+0x4e7483)
    #4 0x6aeb48 (/home/fd/git/JtR/run/john+0x6aeb48)
    #5 0x6b040f (/home/fd/git/JtR/run/john+0x6b040f)
    #6 0x762978 (/home/fd/git/JtR/run/john+0x762978)
    #7 0x7644f5 (/home/fd/git/JtR/run/john+0x7644f5)
    #8 0x7646c0 (/home/fd/git/JtR/run/john+0x7646c0)
    #9 0x7b18da (/home/fd/git/JtR/run/john+0x7b18da)
    #10 0x777853 (/home/fd/git/JtR/run/john+0x777853)
    #11 0x7785b9 (/home/fd/git/JtR/run/john+0x7785b9)
    #12 0x38dd621d64 (/usr/lib64/libc-2.18.so+0x21d64)
    #13 0x4065b4 (/home/fd/git/JtR/run/john+0x4065b4)
llvm-symbolizer: for the -functions option: 'short' is invalid value for boolean argument! Try 0 or 1
    #0 0x4ee9b6 in rarvm_getbits /home/fd/git/JtR/src/unrarvm.c:228
    #1 0x4eeae8 in rarvm_read_data /home/fd/git/JtR/src/unrarvm.c:265
    #2 0x4e24d4 in add_vm_code /home/fd/git/JtR/src/unrar.c:615
    #3 0x4e7483 in read_vm_code_PPM /home/fd/git/JtR/src/unrar.c:893
    #4 0x6aeb48 in crypt_all._omp_fn.1 /home/fd/git/JtR/src/rar_fmt_plug.c:801
    #5 0x6b040f in crypt_all /home/fd/git/JtR/src/rar_fmt_plug.c:706
    #6 0x762978 in crk_password_loop /home/fd/git/JtR/src/cracker.c:685
    #7 0x7644f5 in crk_salt_loop /home/fd/git/JtR/src/cracker.c:743
    #8 0x7646c0 in crk_process_key /home/fd/git/JtR/src/cracker.c:780
    #9 0x7b18da in do_wordlist_crack /home/fd/git/JtR/src/wordlist.c:1227 (discriminator 1)
    #10 0x777853 in john_run /home/fd/git/JtR/src/john.c:1358
    #11 0x7785b9 in main /home/fd/git/JtR/src/john.c:1651
    #12 0x38dd621d64 in ?? ??:0
    #13 0x4065b4 in _start ??:?


    #0 0x7f3cf8a74219 (/usr/lib64/libasan.so.0.0.0+0x16219)
    #1 0x792293 (/home/fd/git/JtR/run/john+0x792293)
    #2 0x4de9ea (/home/fd/git/JtR/run/john+0x4de9ea)
    #3 0x4e63bd (/home/fd/git/JtR/run/john+0x4e63bd)
    #4 0x6aeb48 (/home/fd/git/JtR/run/john+0x6aeb48)
    #5 0x6b040f (/home/fd/git/JtR/run/john+0x6b040f)
    #6 0x762978 (/home/fd/git/JtR/run/john+0x762978)
    #7 0x7644f5 (/home/fd/git/JtR/run/john+0x7644f5)
    #8 0x7646c0 (/home/fd/git/JtR/run/john+0x7646c0)
    #9 0x7b18da (/home/fd/git/JtR/run/john+0x7b18da)
    #10 0x777853 (/home/fd/git/JtR/run/john+0x777853)
    #11 0x7785b9 (/home/fd/git/JtR/run/john+0x7785b9)
    #12 0x38dd621d64 (/usr/lib64/libc-2.18.so+0x21d64)
    #0 0x7f3cf8a74219 in malloc ??:?
    #1 0x792293 in mem_alloc_func /home/fd/git/JtR/src/memory.c:72
    #2 0x4de9ea in rar_malloc /home/fd/git/JtR/src/unrarhlp.c:39
    #3 0x4e63bd in read_vm_code_PPM /home/fd/git/JtR/src/unrar.c:880
    #4 0x6aeb48 in crypt_all._omp_fn.1 /home/fd/git/JtR/src/rar_fmt_plug.c:801
    #5 0x6b040f in crypt_all /home/fd/git/JtR/src/rar_fmt_plug.c:706
    #6 0x762978 in crk_password_loop /home/fd/git/JtR/src/cracker.c:685
    #7 0x7644f5 in crk_salt_loop /home/fd/git/JtR/src/cracker.c:743
    #8 0x7646c0 in crk_process_key /home/fd/git/JtR/src/cracker.c:780
    #9 0x7b18da in do_wordlist_crack /home/fd/git/JtR/src/wordlist.c:1227 (discriminator 1)
    #10 0x777853 in john_run /home/fd/git/JtR/src/john.c:1358
    #11 0x7785b9 in main /home/fd/git/JtR/src/john.c:1651
    #12 0x38dd621d64 in ?? ??:0
```

**Comment by magnumripper** [comment]
We could look for updates in clamav unrar before doing anything else but their code is not used as-is.

**Comment by magnumripper** [comment]
I synced from upstream but that did not fix the issue.

**Comment by magnumripper** [comment]
It's kinda annoying that GitHub auto-closes the issue from the phrase "This did not fix #1000."

**Comment by jfoug** [comment]
I think I have a 'proper' fix for this now:  I have a line that prints 'out of data' when the problem is detected.  Here is what I now see (after rolling back to alloc == size + 2)

```
12345            (u25-rar)
12345            (u25-rar)
d�ffff�ff�fff    (u21-rar)
d�ffff�ff�fff    (u21-rar)
d�ffff�ff�fff    (u21-rar)
Bert$ErnIE       (u3-rar)
Bert$ErnIE       (u3-rar)
Bert$ErnIE       (u3-rar)
out of data
out of data
out of data
out of data
out of data
Bert$ErnIE       (u3-rar3hp)
Bert$ErnIE       (u3-rar3hp)
Bert$ErnIE       (u3-rar)
```

No ASAN, and no crash.  The only change is in rarvm_getbits() function.  I will get that checked in shortly.

**Comment by jfoug** [comment]
This aa9589e should be proper fix for both #1000 and #1001
